data_structures/src/mainnet_validations.rs

```diff
@@ -16,6 +18,145 @@ pub const FIRST_HARD_FORK: Epoch = 192000;
 /// 28 April 2021 @ 9:00:00 UTC
 pub const SECOND_HARD_FORK: Epoch = 376320;
 
+/// TAPI Engine
+#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
+pub struct TapiEngine {
+    /// bit votes counter by bits
+    pub bit_tapi_counter: BitTapiCounter,
+    /// wip activation
+    pub wip_activation: HashMap<String, Epoch>,
+}
+
+impl TapiEngine {
+    pub fn update_bit_counter(&mut self, v: u32, epoch: Epoch, avoid_wip_list: &HashSet<String>) {
+        // In case of empty epochs, they would be considered as blocks with tapi version to 0
+        // In order to not update bit counter from old blocks where the block version was not used,
+        // the first time (bit_tapi_counter.last_epoch == 0) would be skipped in this conditional branch
+        if self.bit_tapi_counter.last_epoch != 0 && epoch > self.bit_tapi_counter.last_epoch + 1 {
+            let init = self.bit_tapi_counter.last_epoch + 1;
+            let end = epoch;
+            for i in init..end {
+                self.update_bit_counter(0, i, avoid_wip_list);
+            }
+        }
+        for n in 0..32 {
```

Comment:
Remove hardcoded number:

```suggestion
        for n in 0..self.bit_tapi_counter.info.len() {
```

I was thinking that it would be better to use an iterator, like

    for (n, bit_counter) in self.bit_tapi_counter.iter_mut(epoch) {

But it is tricky to implement that, and the benefit is small.